ocaml/xcp-rrdd/bin/rrdp-dcmi/rrdp_dcmi.ml

```
@@ -0,0 +1,88 @@
+(*
+ * Copyright (c) 2024 Cloud Software Group, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+(** Read power measurements from IPMI DCMI where available.
+    There is also IPMI SDR entity 21 that returns the same information (power consumption in watts),
+    but isn't always available, and seems to be slower to read, especially when missing.
+ *)
+
+open Rrdd_plugin
+
+module Process = Process (struct let name = "xcp-rrdd-dcmi" end)
+
+open Process
+
+let ipmitool_bin = "/usr/bin/ipmitool"
+
+let ipmitool args =
+  (* we connect to the local /dev/ipmi0 if available to read measurements from local BMC *)
+  ipmitool_bin :: "-I" :: "open" :: args |> String.concat " "
+
+let discover () =
+  Utils.exec_cmd
+    (module Process.D)
+    ~cmdstring:(ipmitool ["dcmi"; "discover"])
+    ~f:(fun line ->
+      D.debug "DCMI discover: %s" line ;
+      if String.trim line = "Power management available" then
+        Some true
+      else
+        None
+    )
+
+let ( let* ) = Option.bind
+
+let get_dcmi_power_reading () =
+  Utils.exec_cmd
+    (module Process.D)
+    ~cmdstring:(ipmitool ["dcmi"; "power"; "reading"])
+    ~f:(fun line ->
+      let* k, v = line |> Astring.String.cut ~sep:":" in
```

Would it be easier to read the expected output using Sscanf?

---

If I use scanning indications and ignore exceptions from Scanf then this format string would work: `" %s@: %f %s"`